platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/isolation/IsolatedActionSerializer.kt

```diff
@@ -0,0 +1,186 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.isolation
+
+import org.gradle.api.IsolatedAction
+import org.gradle.configurationcache.ConfigurationCacheError
+import org.gradle.configurationcache.extensions.invert
+import org.gradle.configurationcache.extensions.uncheckedCast
+import org.gradle.configurationcache.extensions.useToRun
+import org.gradle.configurationcache.logger
+import org.gradle.configurationcache.problems.ProblemsListener
+import org.gradle.configurationcache.problems.PropertyProblem
+import org.gradle.configurationcache.serialization.ClassDecoder
+import org.gradle.configurationcache.serialization.ClassEncoder
+import org.gradle.configurationcache.serialization.DefaultReadContext
+import org.gradle.configurationcache.serialization.DefaultWriteContext
+import org.gradle.configurationcache.serialization.IsolateOwner
+import org.gradle.configurationcache.serialization.ReadContext
+import org.gradle.configurationcache.serialization.WriteContext
+import org.gradle.configurationcache.serialization.beans.BeanStateReaderLookup
+import org.gradle.configurationcache.serialization.beans.BeanStateWriterLookup
+import org.gradle.configurationcache.serialization.readNonNull
+import org.gradle.configurationcache.serialization.runReadOperation
+import org.gradle.configurationcache.serialization.runWriteOperation
+import org.gradle.configurationcache.serialization.withIsolate
+import org.gradle.configurationcache.services.IsolatedActionCodecsFactory
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.util.IdentityHashMap
+
+
+/**
+ * Serialized state of an object graph containing one or more [IsolatedAction]s.
+ *
+ * @param G type of the root object stored in [graph]
+ */
+class SerializedIsolatedActionGraph<G>(
+    /**
+     * The serialized graph.
+     */
+    val graph: ByteArray,
+
+    /**
+     * External references that are not serialized directly as part of the [graph].
+     * These might include references to classes, value sources and build services.
+     *
+     * Maps the integer written to the serialized [graph] to the external reference.
+     * See [EnvironmentEncoder] and [EnvironmentDecoder] for details.
+     */
+    val environment: Map<Int, Any>,
+)
+
+
+internal
+class IsolatedActionSerializer(
+    private val owner: IsolateOwner,
+    private val beanStateWriterLookup: BeanStateWriterLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun <G : Any> serialize(action: G): SerializedIsolatedActionGraph<G> {
```

Comment:
I noticed we can serialize anything here. Maybe some comment telling a passerby what guarantees/requirements are provided/demanded from this method

Reply:
Maybe. I think the class name and surrounding types are enough to get to `IsolatedAction` and its requirements.